File: object-oriented-programming/virtual-world-python/animals/animal.py
from abc import abstractmethod
from organism import Organism
import random

class Animal(Organism) :

    def __init__(self, species, x, y, strength, initiative, world):
        super().__init__(species, x, y, strength, initiative, world, True, 0)

    
    def action(self) :
        if not self._can_breed :
            self._breed_cooldown-=1
            if self._breed_cooldown == 0 :
                self._can_breed = True

        dy = [1, 0, -1, 0]
        dx = [0, 1, 0, -1]
        for i in range(4) :
            direction = random.randint(0, 3)
            newX = self._x + dx[direction]
            newY = self._y + dy[direction]
            if not self._world.in_bounds(newX, newY) : continue
            self.move_organism(newX, newY)
            self._world.report(f"{self.species.display_name()} moved!")
            return
    
    def collision(self, other) :
        dy = [1, 0, -1, 0]
        dx = [0, 1, 0, -1]
        if self.species == other.species :
            for i in range(4) :
                newY = self.y + dy[i]
                newX = self.x + dx[i]
                if (self._world.is_free(newX, newY) and self.check_can_breed() and other.check_can_breed()) :
                    self.breed(newX, newY)
                    self.set_breed_cooldown()
                    other.set_breed_cooldown()
                    self._world.report(f"{self._species.display_name()} offspring appeared!")
                    break
        elif(self.strength > other.strength) :
            self._world.report(f"{self.species.display_name()} killed {other.species.display_name()}!")
            other.mark_dead()
            self._world.set_organism(None, self.x, self.y)
            self._y = other.y
            self._x = other.x
            self._world.set_organism(self, self.x, self.y)
        else :
            self._world.report(other.species.display_name() + " killed " + self.species.display_name() + "!")
            self.mark_dead()
            self._world.set_organism(None, self.x, self.y)
            other.move_organism(self.x, self.y)

    @abstractmethod
    def breed(self, x, y) :
        pass
    

File: object-oriented-programming/virtual-world-java/src/Game/Model/Animals/Animal.java
package Game.Model.Animals;

import Game.Model.Organism;
import Game.Model.Species;
import Game.Model.World;

import java.util.Random;

public abstract class Animal extends Organism {
    Random random = new Random();

    public Animal(Species species, int x, int y, int strength, int initiative, World world, int breedCooldown, boolean canBreed){
        super(species, x, y, strength, initiative, world, breedCooldown, canBreed);
    }

    public void action(){
        if(!canBreed){
            --breedCooldown;
            if(breedCooldown == 0){
                canBreed = true;
            }
        }
        for(int i = 0; i < 4; i++){
            int[] dy = {1, 0, -1, 0};
            int[] dx = {0, 1, 0, -1};
            int direction = random.nextInt(4);
            int newX = x + dx[direction];
            int newY = y + dy[direction];
            if(!world.inBounds(newY, newX)) continue;
            moveOrganism(newY, newX);
            world.report(species.displayName() + " moved!");
            return;
        }
    }

    public void collision(Organism other){
        if(getSpecies() == other.getSpecies()){
            int[] dy = {1, 0, -1, 0};
            int[] dx = {0, 1, 0, -1};
            int newY, newX;
            for(int i = 0; i < 4; i++) {
                newY = y + dy[i];
                newX = x + dx[i];
                if (world.isFree(newY, newX) && this.checkCanBreed() && other.checkCanBreed()) {
                    breed(newY, newX);
                    this.setBreedCooldown();
                    other.setBreedCooldown();
                    world.report(species.displayName() + " offspring appeared!");
                    break;
                }
            }
        }

        else if(getStrength() > other.getStrength()){
            world.report(species.displayName() + " killed " + other.getSpecies().displayName() + "!");
            other.markDead();
            world.setOrganism(null, y, x);
            y = other.getY();
            x = other.getX();
            world.setOrganism(this, y, x);
        }

        else {
            world.report(other.getSpecies().displayName() + " killed " + species.displayName() + "!");
            markDead();
            world.setOrganism(null, y, x);
            other.moveOrganism(y, x);
        }
    }

    public abstract Animal breed(int y, int x);
}
